firebase-firestore/src/main/java/com/google/firebase/firestore/core/CompositeFilter.java

```diff
@@ -0,0 +1,347 @@
+   * FieldFilters (i.e. it does not contain any other composite filters).
+   */
+  public boolean isFlatAndFilter() {
+    return isAnd() && !containsCompositeFilters();
+  }
+
+  /**
+   * Returns true if this composite filter contains any other composite filter. Returns false
+   * otherwise.
+   */
+  public boolean containsCompositeFilters() {
+    for (Filter filter : filters) {
+      if (filter instanceof CompositeFilter) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Performs a depth-first search to find and return the first FieldFilter in the composite filter
+   * that satisfies the condition. Returns null if none of the FieldFilters satisfy the condition.
+   */
+  public FieldFilter firstFieldFilterWhere(FieldFilterCondition condition) {
+    for (Filter filter : filters) {
+      if (filter instanceof FieldFilter && condition.check(((FieldFilter) filter))) {
+        return (FieldFilter) filter;
+      } else if (filter instanceof CompositeFilter) {
+        FieldFilter found = ((CompositeFilter) filter).firstFieldFilterWhere(condition);
+        if (found != null) {
+          return found;
+        }
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Returns the first inequality filter contained within this composite filter. Returns null if it
+   * does not contain any inequalities.
+   */
+  public FieldFilter getInequalityFilter() {
+    return firstFieldFilterWhere(f -> f.isInequality());
+  }
+
+  @Override
+  public boolean matches(Document doc) {
+    if (isAnd()) {
+      // For conjunctions, all filters must match, so return false if any filter doesn't match.
+      for (Filter filter : filters) {
+        if (!filter.matches(doc)) {
+          return false;
+        }
+      }
+      return true;
+    } else {
+      // For disjunctions, at least one filter should match.
+      for (Filter filter : filters) {
+        if (filter.matches(doc)) {
+          return true;
+        }
+      }
+      return false;
+    }
+  }
+
+  @Override
+  public String getCanonicalId() {
+    StringBuilder builder = new StringBuilder();
+    for (Filter filter : filters) {
+      if (builder.length() == 0) {
+        builder.append(isAnd() ? "and(" : "or(");
+      } else {
+        builder.append(",");
+      }
+      builder.append(filter.getCanonicalId());
+    }
+    builder.append(")");
+    return builder.toString();
+  }
+
+  @Override
+  public String toString() {
+    return getCanonicalId();
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (o == null || !(o instanceof CompositeFilter)) {
+      return false;
+    }
+    CompositeFilter other = (CompositeFilter) o;
+    // Note: This comparison requires order of filters in the list to be the same, and it does not
+    // remove duplicate subfilters from each composite filter. It is therefore way less expensive.
+    return operator == other.operator && filters.equals(other.filters);
```

Comment:
This is interesting and has some wide-reaching implications:

`where(or(filterA, filterB))` is now different from `where(or(filterB, filterA))`. This means that we will now create two different targets for these queries, which adds additional costs for the customer. It would be nice if we could get the canonicalId and the `equals()` to match in this case.

Reply:
We should discuss this